src/promptflow/promptflow/_sdk/_service/apis/collector.py

```diff
@@ -60,21 +67,29 @@ def trace_collector(get_created_by_info_with_cache: Callable, logger: logging.Lo
                 for span in scope_span.spans:
                     # TODO: persist with batch
                     span = Span._from_protobuf_object(span, resource=resource)
-                    span._persist()
+                    if not cloud_trace_only:
+                        span._persist()
                     all_spans.append(span)
 
-        # Create a new thread to write trace to cosmosdb to avoid blocking the main thread
-        ThreadWithContextVars(
-            target=_try_write_trace_to_cosmosdb, args=(all_spans, get_created_by_info_with_cache, logger)
-        ).start()
+        if cloud_trace_only:
+            # If we only trace to cloud, we should make sure the data writing is success before return.
+            _try_write_trace_to_cosmosdb(all_spans, get_created_by_info_with_cache, logger, credential)
```

Comment:
Do we want to raise exception here if meet any exception? Just raise out a problem, should be in another PR.

Reply:
For runtime implement, I think we can throw exception and handle it at runtime http handler.